cassandra/src/main/java/org/hibernate/ogm/datastore/cassandra/CassandraDialect.java

```diff
@@ -183,14 +200,23 @@ public void insertOrUpdateTuple(EntityKey key, Tuple tuple, TupleContext tupleCo
 
 			// insert and update are both 'upsert' in cassandra.
 			Insert insert = insertInto( quote( key.getTable() ) );
-			Object[] columnValues = new Object[updateOps.size()];
+			List<Object> columnValues = new LinkedList<>();
+			Set<String> seenColNames = new HashSet<>();
 			for ( int i = 0; i < updateOps.size(); i++ ) {
 				TupleOperation op = updateOps.get( i );
 				insert.value( quote( op.getColumn() ), QueryBuilder.bindMarker() );
-				columnValues[i] = op.getValue();
+				columnValues.add( op.getValue() );
+				seenColNames.add( op.getColumn() );
+			}
+			for ( int j = 0; j < key.getColumnNames().length; j++ ) {
```

Comment:
- test failures: no clue, but at least we're building up a good picture of where the trace/debug logging statements need to be added ;-)   Please dump the query string from CassandraDialect.bindAndExecute and the DDL from CassandraDatastoreProvider createColumnFamilyIfNeeded and  createSecondaryIndexIfNeeded, that should allow me to figure out what differs between your setup and mine.

mutations need to specify all the *key* columns, or you wind up mutating more rows that you want.